### PR TITLE
Matching algorithm update

### PR DESCRIPTION
**Pull request by Iyanuoluwa-abolade**
- Updated the implementation of Gale-Shapley Algorithm.
- Implemented an API for calculating similarity.
- Implemented cosine similarity when ranks are equal.
- Created a profile form for user that updates in the database.
- Enable users to enter their rank or preference, updating the database accordingly.
- Implemented profile picture as a stretch feature

Image 7-24-24 at 3 14 AM
Image 7-24-24 at 3 11 AM
Image 7-24-24 at 3 13 AM
Image 7-24-24 at 3 15 AM

### CODE REVIEW

File: backend/matchAlgorithm.js [hunk 1]
```diff
@@ -1,40 +1,143 @@
...
+import { PrismaClient } from '@prisma/client';
+import { fetchCosineSimilarity } from './TextSimilarity.js';
 
+const prisma = new PrismaClient();
 
-async function galeShapley() {
+export async function galeShapley(loggedInUserId, userRole) {
   const adopters = await prisma.adopter.findMany();
   const adoptees = await prisma.adoptee.findMany();
+  const rawAdopterPreferences = await prisma.adopterPreference.findMany();
```
Comment (Iyanuoluwa-abolade) [inline backend/matchAlgorithm.js:9]: This fetch all adopters and adoptees including their preferences(rank) from the database

File: backend/matchAlgorithm.js [hunk 2]
```diff
@@ -1,40 +1,143 @@
...
+      .filter(preference => preference.adopterId === adopterId)
+      .sort((a, b) => b.rank - a.rank)
+      .map(preference => preference.adopteeId);
   }
 
   for (const adoptee of adoptees) {
-    const preferences = adopteePreferences.filter((preference) => preference.adopteeId === adoptee.id);
-    for (const preference of preferences) {
-      const adopter = adopters.find((adopter) => adopter.id === preference.adopterId);
-      if (adopter && !matches[adopter.id]) {
```
Comment (Iyanuoluwa-abolade) [inline backend/matchAlgorithm.js]: Each Adopter's ID maps to an array of adoptee IDs which is sorted by preference in a descending order and vice versa

File: backend/matchAlgorithm.js [hunk 3]
```diff
@@ -1,40 +1,143 @@
...
-        matches[adopter.id] = adoptee;
-        break;
+    const adopteeId = adoptee.id;
+    adopteePreferences[adopteeId] = rawAdopteePreferences
+      .filter(preference => preference.adopteeId === adopteeId)
+      .sort((a, b) => b.rank - a.rank)
+      .map(preference => preference.adopterId);
+  }
```
Comment (Iyanuoluwa-abolade) [inline backend/matchAlgorithm.js]: Each Adopter's ID maps to the sorted list of adoptee IDs based on rank in a descending order and vice versa

File: backend/matchAlgorithm.js [hunk 4]
```diff
@@ -1,40 +1,143 @@
...
+            freeUsers.delete(userId);
+            freeUsers.add(currentUserId);
+          }
+        }
       }
     }
+
+    return { userMatches, otherUserMatches };
+  };
+
```
Comment (Iyanuoluwa-abolade) [inline backend/matchAlgorithm.js]: This is the implementation of Gale-Shapley's algorithm

File: backend/matchAlgorithm.js [hunk 5]
```diff
@@ -1,40 +1,143 @@
...
+        update: { score: similarity.similarity },
+        create: { adopterId: parseInt(currentUser.id), adopteeId: parseInt(id), score: similarity.similarity }
+      });
+    }
+
+    return userType === 'Adopter'
+      ? { adopter: currentUser, adoptee: adoptees.find(adoptee => adoptee.id === bestMatch) }
+      : { adopter: adopters.find(adopter => adopter.id === bestMatch), adoptee: currentUser };
+  };
+
```
Comment (Iyanuoluwa-abolade) [inline backend/matchAlgorithm.js:89]: If the ranks are equal, cosine similarity is used to determine the best match based on the highest cosine similarity score

File: backend/matchAlgorithm.js [hunk 6]
```diff
@@ -1,40 +1,143 @@
...
+      const { userMatches, otherUserMatches } = applyGaleShapley(adopterPreferences, adopteePreferences, freeAdopters, proposerIndex);
+
+      return {
+        adopter: currentAdopter,
+        adoptee: adoptees.find(adoptee => adoptee.id === userMatches[currentAdopter.id])
+      };
+    } else {
+      return await applyCosineSimilarity(currentAdopter, currentAdopterPrefList, 'Adopter');
+    }
   }
```
Comment (Iyanuoluwa-abolade) [inline backend/matchAlgorithm.js]: This check if the preferences have different rank or not and the appropriate matching method is used(Gale-Shapley or cosine similarity)

File: backend/matchAlgorithm.js [hunk 7]
```diff
@@ -1,40 +1,143 @@
...
+      return {
+        adopter: adopters.find(adopter => adopter.id === otherUserMatches[currentAdoptee.id]),
+        adoptee: currentAdoptee
+      };
+    } else {
+      return await applyCosineSimilarity(currentAdoptee, currentAdopteePrefList, 'Adoptee');
+    }
+  }
+
+  return { error: "Invalid user role or user not found" };
```
Comment (Iyanuoluwa-abolade) [inline backend/matchAlgorithm.js]: This is similar to adopter role. It checks if the preferences have different rank or not and the appropriate matching method is used(Gale-Shapley or cosine similarity)

**Comment by robintan98** [comment]
This is fantastic -- logic for gale shapley and cosine similarity looks correct to me at a cursory glance (but we need to verify through some testing). Your code style looks great as well, not commented out code, extra whitespaces, etc.

The only things I would suggest for the future are: 1) organize your commits such that one commit implements Gale-Shapley, and another commit should implement cosine similarity; 2) add comments to explain your implementation of both of them.

Approving for now, but again we should test with some test cases before you finish your project.

**Review by robintan98 — APPROVED** [review]

**Review by robintan98 — APPROVED** [review]
Nice!! Great progress. The only thing I would call out is that the size of your commits (i.e. lines of code) are still a bit large. Try to keep them to under <100 in the future.

**Pull request merged**